velox/exec/tests/HashJoinTest.cpp

```diff
@@ -69,163 +95,327 @@ class HashJoinTest : public HiveConnectorTestBase {
-      const std::string& filter,
-      const std::vector<std::string>& outputLayout,
-      core::JoinType joinType) {
-    CursorParameters params;
-    auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
-    params.planNode = PlanBuilder(planNodeIdGenerator)
-                          .values(leftBatches, isLeftParallelizable)
-                          .hashJoin(
-                              leftKeys,
-                              rightKeys,
-                              PlanBuilder(planNodeIdGenerator)
-                                  .values(rightBatches, isRightParallelizable)
-                                  .planNode(),
-                              filter,
-                              outputLayout,
-                              joinType)
-                          .planNode();
-    params.maxDrivers = numDrivers_;
-
-    auto task = ::assertQuery(
-        params, [](auto*) {}, referenceQuery, duckDbQueryRunner_);
-    // A quick sanity check for memory usage reporting. Check that peak total
-    // memory usage for the hash join node is > 0.
-    auto planStats = toPlanStats(task->taskStats());
-    auto joinNodeId = params.planNode->id();
-    ASSERT_TRUE(planStats.at(joinNodeId).peakMemoryBytes > 0);
+  VectorFuzzer getFuzzer() {
+    return VectorFuzzer(fuzzerOpts_, pool_.get());
   }
 
-  void testJoin(
-      const std::vector<TypePtr>& keyTypes,
-      int leftBatchSize,
-      int numLeftBatches,
-      int rightBatchSize,
-      int numRightBatches,
-      const std::string& referenceQuery,
-      const std::string& filter = "",
-      const std::vector<std::string>& outputLayout = {},
-      core::JoinType joinType = core::JoinType::kInner) {
-    auto leftType = makeRowType(keyTypes, "t_");
-    auto rightType = makeRowType(keyTypes, "u_");
-
-    std::vector<RowVectorPtr> leftBatches;
-    std::vector<RowVectorPtr> rightBatches;
-    auto opts = fuzzerOpts_;
-    if (leftBatchSize != 0) {
-      opts.vectorSize = leftBatchSize;
-      fuzzer_.setOptions(opts);
-      for (int i = 0; i < numLeftBatches; ++i) {
-        leftBatches.push_back(fuzzer_.fuzzRow(leftType));
+  // Make batches with random data.
+  std::vector<RowVectorPtr> makeBatches(
+      int32_t batchSize,
+      int32_t numBatches,
+      RowTypePtr rowType,
+      double nullRatio = 0.1) {
+    std::vector<RowVectorPtr> batches;
+    batches.reserve(numBatches);
+    if (batchSize != 0) {
+      fuzzerOpts_.vectorSize = batchSize;
+      fuzzerOpts_.nullRatio = nullRatio;
+      VectorFuzzer fuzzer = getFuzzer();
+      for (int32_t i = 0; i < numBatches; ++i) {
+        batches.push_back(fuzzer.fuzzRow(rowType));
       }
     } else {
-      for (int i = 0; i < numLeftBatches; ++i) {
-        leftBatches.push_back(RowVector::createEmpty(leftType, pool_.get()));
+      batches.push_back(RowVector::createEmpty(rowType, pool_.get()));
+    }
+    // NOTE: we generate a number of vectors with a fresh new fuzzer init with
```

Comment:
Does it mean we don't test the case where some records do not have a match?

Reply:
If the number of probe and build vectors doesn't match, then there will be some don't match. I think most test cases we choose different number of probe and build vectors.